Refactor with strands

statefb opened the pull request:
*Issue #, if available:*
close #897 

*Description of changes:*
To introduce MCP support (#866), refactor to replace current implementation with [strands SDK](https://github.com/strands-agents/sdk-python)

### Tool Implementation
- Refactored Strands tools to be thin wrappers around existing implementations
- knowledge_tool_strands.py: Uses existing search_knowledge and KnowledgeToolInput
- bedrock_agent_tool_strands.py: Uses existing _bedrock_agent_invoke and BedrockAgentInput
- internet_search_tool_strands.py: Already implemented with Firecrawl → DuckDuckGo fallback
- Utilizes Strands tool execution hooks (`BeforeToolInvocationEvent` / `AfterToolInvocationEvent`) to include `source_id` for citation support. 
  - Note: Uses [strands.experimental.hooks](https://strandsagents.com/latest/documentation/docs/api-reference/experimental/) module.

### Chat Routing System
- Split `chat()` into common and framework specific implementation.
  1. (Common preprocess) Setup messages and LLM parameters.
  2. (Framework specific) Converse with LLM and build response messages.
    - `USE_STRANDS == true` → Routes to `converse_with_strands()` (Strands implementation)
    - `USE_STRANDS == false` → Routes to `converse_legacy()` (original implementation with deprecation warning)
  3. (Common postprocess) Update conversation and store into database.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

statefb committed:
wip: add strands integration core files

statefb committed:
update deps

statefb committed:
wip

statefb committed:
update firecrawl-py version

statefb committed:
insert debug logs

statefb committed:
fix: on_stop impl

statefb committed:
wip

statefb committed:
Fix reasoning content extraction from Strands AgentResult

- Update _extract_reasoning_content_from_agent_result to handle Strands' actual reasoning structure
- Change from looking for 'type' == 'reasoning' to 'reasoningContent' key
- Extract reasoning text from reasoningContent.reasoningText.text path
- Use actual signature from Strands instead of hardcoded value
- Ensure reasoning content is properly added to thinking_log for persistence
- Align with chat_legacy logic for reasoning content handling

statefb committed:
fix: tool use id conversion

statefb committed:
fix: internet

statefb committed:
add debug log on websocket.py

statefb committed:
add pytest

statefb committed:
add debug log on usePostMessageStreaming

statefb committed:
fix: tool input / output not displayed

statefb committed:
fix: reasoning not persist

statefb committed:
add calc tool for testing

statefb committed:
fix: multi turn conversation

statefb committed:
fix: tool registry

statefb committed:
fix: wait complete tool input

statefb committed:
fix: citation

statefb committed:
fix: tool registry

statefb committed:
fix: tool input consistency

statefb committed:
fix: support list

statefb committed:
fix: list citation

statefb committed:
fix: citation

statefb committed:
remove context

statefb committed:
fix: knowledge tool strands to return list

statefb committed:
refactor

statefb committed:
update strands version

statefb committed:
wip: chat strands refactor

statefb committed:
refactor: call back handler

statefb committed:
add post processing

statefb committed:
fix: tools / utils

statefb committed:
fix: attatchment docs

statefb committed:
fix: image content

statefb committed:
fix: continue generation

statefb committed:
fix: Skip instruction messages as they are handled separately via message_map

statefb committed:
change log level for websocket.py

statefb committed:
lint and add comment

statefb committed:
remove deprecated refactorings

statefb committed:
fix: unittest

statefb committed:
fix tools to return result as strands formats

statefb committed:
rename modules

statefb committed:
fix: skip reasoning / tool content to construct strunds message before invoke LLM

statefb committed:
fix: knowledge_search

statefb committed:
fix: source id citation

statefb committed:
add: prompt cache (system, tool)

statefb committed:
add message cache

statefb committed:
insert debug log

statefb committed:
return empty list when no tool

statefb committed:
fix: tool util

statefb committed:
fix bedrock agent tool

statefb committed:
add deprecated decorator

statefb committed:
update documents including examples

statefb committed:
add deprication decorator

statefb committed:
switch fetch_available_agent_tools for strands

statefb committed:
refactor modules for readability

statefb committed:
chore: mypy

statefb committed:
refactor: simplify on_stop lambda in process_chat_input

Yukinobu-Mine commented on backend/app/routes/bot.py:
```diff
@@ -21,6 +21,7 @@
     GenerationParams,
     Knowledge,
     PlainTool,
```
`PlainTool` の import は削除できそうです。

statefb replied:
対応済み：https://github.com/aws-samples/bedrock-chat/pull/902/commits/61fd258bdad609d06bd2a58a033658e56f1e91b5

Yukinobu-Mine commented on backend/app/strands_integration/handlers/callback_handler.py:
```diff
@@ -0,0 +1,61 @@
+            self.collected_reasoning.append(reasoning_text)
+        elif "thinking" in kwargs and self.on_reasoning:
+            thinking_text = kwargs.get("thinking", "")
+            self.on_reasoning(thinking_text)
```
こちらで呼び出すのは `on_thinking()` ではなく `on_reasoning()` で合ってますか?

statefb replied:
`data`および`reasoning`以外は削除: https://github.com/aws-samples/bedrock-chat/pull/902/commits/54b5398afa0c1f6ccdb1743162f72f261392f103

statefb committed:
remove unused imports on routes/bot.py

statefb committed:
fix: support legacy for `fetch_available_agent_tools`

statefb committed:
chore: lint

statefb committed:
remove unused tests

statefb committed:
feat: implement telemetry management and data extraction for Strands integration

Yukinobu-Mine commented on backend/app/strands_integration/telemetry/telemetry_manager.py:
```diff
@@ -0,0 +1,32 @@
+    def setup(self, conversation_id: str, user_id: str):
+        """Setup telemetry with custom span processors."""
+        # Setup console exporter for development
+        self.telemetry.setup_console_exporter()
```
for development とありますが、Console Exporter は常に有効化する方針ですか?
それとも、今後のコミットで削除予定ですか?

statefb replied:
未使用です。削除します

statefb replied:
https://github.com/aws-samples/bedrock-chat/pull/902/commits/dac05895abf6b13d7c3df0c962a638c71b0784d2

statefb committed:
convert relative import to absolute

statefb committed:
refactor: reorganize imports and remove console exporter setup

statefb committed:
Merge branch 'v3' into refactor-strands

Yukinobu-Mine commented on backend/app/strands_integration/tools/knowledge_search.py:
```diff
@@ -0,0 +1,117 @@
+    """Create a knowledge search tool with bot context captured in closure."""
+
+    @tool
+    def knowledge_search(query: str) -> dict:
```
ツール名が関数名と同じ `knowledge_search` で登録されると思いますが、フロントエンド側では[従来通り](https://github.com/aws-samples/bedrock-chat/blob/3811ec523b20ea644c3dc7b9bdd9b87515468158/backend/app/agents/tools/knowledge.py#L57)の `knowledge_base_tool` というツール名が[期待されている](https://github.com/aws-samples/bedrock-chat/blob/3811ec523b20ea644c3dc7b9bdd9b87515468158/frontend/src/i18n/en/index.ts#L189)と思います。
strands 利用時におけるツール名の変化は意図されていますか?

statefb replied:
同じく対応済み: https://github.com/aws-samples/bedrock-chat/pull/902/commits/2a81333c30fda654fe4d514003a01e7390276427

Yukinobu-Mine commented on backend/app/strands_integration/tools/bedrock_agent.py:
```diff
@@ -0,0 +1,318 @@
+    """Create a Bedrock Agent tool with bot context captured in closure."""
+
+    @tool
+    def bedrock_agent_invoke(query: str) -> dict:
```
ツール名が関数名と同じ `bedrock_agent_invoke` で登録されると思いますが、フロントエンド側では[従来通り](https://github.com/aws-samples/bedrock-chat/blob/3811ec523b20ea644c3dc7b9bdd9b87515468158/backend/app/agents/tools/bedrock_agent.py#L225)の `bedrock_agent` というツール名が[期待されている](https://github.com/aws-samples/bedrock-chat/blob/3811ec523b20ea644c3dc7b9bdd9b87515468158/frontend/src/i18n/en/index.ts#L193)と思います。
strands 利用時におけるツール名の変化は意図されていますか?

statefb replied:
ご指摘ありがとうございます。i18nの整合性を考慮できていなかったですね。。従来のbedrock_agentに戻すことに問題はないので、戻します

statefb replied:
対応済み: https://github.com/aws-samples/bedrock-chat/pull/902/commits/1c25b624a91e276547e66fe1fe6e7234710b1959

statefb committed:
add type notation

statefb committed:
fix: change back `bedrock_agent_invoke` to original: `bedrock_agent`

statefb committed:
fix: change back `knowledge_search` to original: `knowledge_base_tool`

statefb committed:
chore: lint

statefb committed:
fix: tool example

statefb committed:
add RAG support for model which does not support tool

Yukinobu-Mine committed:
Refactor implementation of strands agents migration.
- Split `chat()` into common and framework specific implementation.
  1. (Common preprocess) Setup messages and LLM parameters.
  2. (Framework specific) Converse with LLM and build response messages.
  3. (Common postprocess) Update conversation and store into database.

- Consolidate the code for converting between application-defined models (messages, contents and tools) and Strands models into `converters` module.

- Fix some bugs in streaming processing
  - Fail to continue generation based on a message with `ReasoningContent`.
  - Not handling `redactedContent` correctly.

statefb commented:
Memo: reflect #927 as well

Yukinobu-Mine committed:
Merge branch 'v3' into refactor-strands

Yukinobu-Mine committed:
Merge branch 'v3' into refactor-strands

HirokiNakatani commented:
When is it scheduled to be released?
If MCP can be used, the range of support will be greatly expanded.

statefb committed:
refactor: add _ prefix to local helper functions in strands converters

statefb committed:
Merge branch 'v3' into refactor-strands

statefb committed:
chore: lint

HirokiNakatani commented:
Thank you for your response. We would appreciate your cooperation in supporting the MCP server.
